File: 1644-maximum-number-of-non-overlapping-substrings/maximum-number-of-non-overlapping-substrings.java
import java.util.*;

class Solution {
    public List<String> maxNumOfSubstrings(String s) {
        int n = s.length();
        int[] first = new int[26];
        int[] last = new int[26];
        Arrays.fill(first, -1);
        Arrays.fill(last, -1);

        for (int i = 0; i < n; i++) {
            int ch = s.charAt(i) - 'a';
            if (first[ch] == -1) first[ch] = i;
            last[ch] = i;
        }

        // Collect all valid expanded ranges [left, right]
        List<int[]> ranges = new ArrayList<>();
        for (int c = 0; c < 26; c++) {
            if (first[c] == -1) continue;
            
            int left = first[c];
            int right = last[c];
            boolean valid = true;

            for (int i = left; i <= right; i++) {
                int ch = s.charAt(i) - 'a';
                if (first[ch] < left) {
                    valid = false; // Cannot start at left because a character extends further left
                    break;
                }
                right = Math.max(right, last[ch]);
            }

            if (valid) {
                ranges.add(new int[]{left, right});
            }
        }

        // Sort ranges by their end position (right boundary)
        ranges.sort((a, b) -> Integer.compare(a[1], b[1]));

        // Greedy interval selection
        List<String> result = new ArrayList<>();
        int lastEnd = -1;

        for (int[] range : ranges) {
            int left = range[0];
            int right = range[1];
            if (left > lastEnd) {
                result.add(s.substring(left, right + 1));
                lastEnd = right;
            }
        }

        return result;
    }
}

File: 1644-maximum-number-of-non-overlapping-substrings/maximum-number-of-non-overlapping-substrings.cpp
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<string> maxNumOfSubstrings(string s) {
        int n = s.length();
        vector<int> first(26, -1);
        vector<int> last(26, -1);

        // Step 1: Find first and last indices for each character
        for (int i = 0; i < n; i++) {
            int ch = s[i] - 'a';
            if (first[ch] == -1) {
                first[ch] = i;
            }
            last[ch] = i;
        }

        // Step 2: Expand ranges for each present character
        vector<pair<int, int>> ranges;
        for (int c = 0; c < 26; c++) {
            if (first[c] == -1) continue;

            int left = first[c];
            int right = last[c];
            bool valid = true;

            for (int i = left; i <= right; i++) {
                int ch = s[i] - 'a';
                if (first[ch] < left) {
                    valid = false; // Cannot start at 'left' without violating character completeness
                    break;
                }
                right = max(right, last[ch]);
            }

            if (valid) {
                ranges.push_back({left, right});
            }
        }

        // Step 3: Sort valid ranges by their right endpoint
        sort(ranges.begin(), ranges.end(), [](const pair<int, int>& a, const pair<int, int>& b) {
            return a.second < b.second;
        });

        // Step 4: Greedy selection of non-overlapping substrings
        vector<string> result;
        int lastEnd = -1;

        for (const auto& r : ranges) {
            int left = r.first;
            int right = r.second;
            if (left > lastEnd) {
                result.push_back(s.substr(left, right - left + 1));
                lastEnd = right;
            }
        }

        return result;
    }
};